### ISSUE TITLE
Upgrading to version 0.58.4 increases the number of database connections by 50%

### ISSUE THREAD

**Issue by Giaco9** [DataDog/dd-trace-py#3303]
### Which version of dd-trace-py are you using?
0.58.4

### Which version of pip are you using?
$ pip --version
pip 20.3.4 from /usr/local/lib/python2.7/site-packages/pip (python 2.7)

_ddtrace requires pip>=18 to install one of our pre-built wheels_

### Which version of the libraries are you using?

```
algoliasearch==2.0.4
amqp==2.6.1
analytics-python==1.2.9
apache-bookkeeper-client==4.12.0
appdirs==1.4.4
asn1crypto==1.4.0
attrs==21.4.0
auth0-python==3.14.0
autopep8==1.6.0
azure-core==1.21.1
azure-nspkg==3.0.2
azure-storage-blob==12.8.1
azure-storage-nspkg==3.1.0
Babel==1.3
backoff==1.8.0
backports.csv==1.0.7
backports.functools-lru-cache==1.6.4
backports.shutil-get-terminal-size==1.0.0
backports.ssl-match-hostname==3.4.0.2
beautifulsoup4==4.3.2
billiard==3.6.4.0
bleach==1.5.0
boto3==1.14.45
botocore==1.17.63
cached-property==1.5.2
cachetools==3.1.1
celery==4.4.7
certifi==2021.10.8
cffi==1.15.0
chardet==3.0.4
configparser==4.0.2
contextlib2==0.6.0.post1
coreapi==2.3.3
coreschema==0.0.4
coverage==4.0
cryptography==2.9.2
cssmin==0.2.0
currencylayer==1.0
ddtrace==0.58.4
decorator==4.4.2
defusedxml==0.4.1
diff-match-patch==20200713
disqus-python==0.4.2
Django==1.11.29
django-admin-sortable2==0.6.4
django-ajax-selects==1.6.1
django-appconf==1.0.4
django-assets==0.11
django-ckeditor==5.6.1
django-codemirror-widget==0.4.0
django-constance==2.5.0
django-cookies-samesite==0.5.1
django-cors-headers==2.4.0
django-crispy-forms==1.7.0
django-debug-panel==0.8.3
django-debug-toolbar==1.11.1
django-environ==0.4.5
django-extensions==1.9.7
django-extra-fields==0.9
django-filter==1.1.0
django-fsm==2.6.0
django-fsm-admin==1.2.4
django-grappelli==2.10.1
django-guardian==1.4.9
django-health-check==3.5.1
django-htmlmin==0.7.0
django-import-export==1.2.0
django-js-asset==1.2.3
django-js-reverse==0.6.1
django-jsoneditor==0.1.2
django-longerusernameandemail==0.5.7
django-memoize==2.0.0
django-model-utils==3.0.0
django-multiselectfield==0.1.8
django-nose==1.4.4
django-oauth-toolkit==1.1.2
django-phonenumber-field==2.2.0
django-picklefield==2.1.1
django-positions==0.6.0
django-redis-cache==2.0.0
django-reversion==2.0.8
django-settings-export==1.0.6
django-silk==3.0.4
django-smtp-ssl==1.0
django-sslserver==0.19
django-staticfiles==1.2.1
django-storages==1.7.1
django-tinymce4==4.2
django-treebeard==4.3.1
django-uuidfield==0.5.0
django-yamlfield==1.0.3
djangorestframework==3.7.7
djangorestframework-jsonapi==2.6.0
djangorestframework-queryfields==1.0.0
dm.xmlsec.binding==1.3.3
docopt==0.6.2
docutils==0.12
drf-yasg==1.14.0
elastic-apm==5.10.1
entrypoints==0.3
enum34==1.1.10
et-xmlfile==1.0.1
expiringdict==1.2.1
extras==1.0.0
Faker==3.0.1
fastavro==0.24.0
fcm-django==0.2.19
feedparser==5.1.3
filestack-python==2.8.1
freezegun==0.3.12
funcsigs==1.0.2
functools32==3.2.3.post2
future==0.18.2
futures==3.3.0
gevent==21.8.0
google-api-python-client==1.7.4
google-auth==2.6.0
google-auth-httplib2==0.0.4
gprof2dot==2021.2.21
greenlet==1.1.2
grpcio==1.27.1
gspread==0.6.2
html5lib==0.9999999
httplib2==0.10.3
hubspot-conn==1.8.0
humanfriendly==4.9
icalendar==4.0.2
idna==2.8
importlib-metadata==2.1.3
inflection==0.3.1
ipaddress==1.0.23
ipdb==0.11
iptools==0.6.1
ipython==5.10.0
ipython-genutils==0.2.0
iso8601==0.1.10
isodate==0.6.0
itsdangerous==0.24
itypes==1.2.0
jdcal==1.4.1
Jinja2==2.7.3
jinja2-django-tags==0.2
jmespath==0.10.0
jsmin==2.0.11
jsonfield==1.0.0
jsonmodels==1.4
jsonpickle==1.4.1
jsonschema==2.5.1
keyring==12.0.1
kombu==4.6.11
ldclient-py==6.9.1
linecache2==1.0.0
logaugment==0.1.3
lxml==4.6.5
Markdown==2.6.4
MarkupPy==1.14
MarkupSafe==0.23
mock==3.0.5
mongoengine==0.9.0
monotonic==1.6
msrest==0.6.21
ndg-httpsclient==0.4.1
ndjson==0.3.1
networkx==2.1
nose==1.3.7
num2words==0.5.10
numpy==1.16.2
oauth2client==3.0.0
oauthlib==3.1.0
odfpy==1.4.1
openpyxl==2.6.4
oscrypto==1.2.1
packaging==20.9
parameterized==0.7.4
parse==1.6.6
pathlib2==2.3.7.post1
pbr==5.8.1
pep562==1.1
pexpect==4.8.0
phonenumberslite==8.10.7
pickleshare==0.7.5
Pillow==5.0.0
prometheus-client==0.12.0
prompt-toolkit==1.0.18
protobuf==3.17.3
psycopg2==2.7.7
ptyprocess==0.7.0
pulsar-client==2.6.1
py==1.11.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
pycodestyle==2.8.0
pycparser==2.21
pydash==3.4.7
pydot==1.4.1
pyfcm==1.4.5
Pygments==2.5.2
PyJWT==1.7.1
pymmh3==0.0.5
pymongo==2.9.5
pyOpenSSL==19.1.0
pyparsing==2.4.7
pyrecord==1.0.1
pyRFC3339==1.1
pyrsistent==0.16.1
python-dateutil==2.7.5
python-http-client==3.3.6
python-mimeparse==1.6.0
python-openid==2.2.5
python-saml==2.4.0
python-slugify==1.1.4
python-social-auth==0.6
pytz==2018.9
PyYAML==5.3.1
ratelimit==2.2.1
recaptcha-client==1.0.6
recurly==2.9.9
redis==3.3.11
requests==2.22.0
requests-mock==0.7.0
requests-oauthlib==1.3.1
requests-toolbelt==0.9.1
retry==0.9.2
rollbar==0.15.2
rsa==3.4
rstview==0.2.1
ruamel.ordereddict==0.4.15
ruamel.yaml==0.16.13
ruamel.yaml.clib==0.2.2
s3fs==0.2.2
s3transfer==0.3.7
scandir==1.10.0
SecretStorage==2.3.1
semver==2.13.0
sendgrid==5.6.0
simpleflake==0.1.2
simplegeneric==0.8.1
simplejson==3.6.5
six==1.15.0
slugify==0.0.1
speaklater==1.3
sqlparse==0.2.0
strict-rfc3339==0.6
tablib==0.14.0
tblib==1.4.0
tenacity==7.0.0
testtools==1.8.0
text-unidecode==1.3
toml==0.10.2
tqdm==4.15.0
traceback2==1.4.0
trafaret==1.2.0
traitlets==4.3.3
typing==3.10.0.0
tzlocal==2.1
ua-parser==0.3.6
undecorated==0.3.0
unicodecsv==0.14.1
Unidecode==1.2.0
unittest2==1.1.0
uritemplate==3.0.1
urllib3==1.24.3
user-agents==0.3.1
uuid==1.30
uWSGI==2.0.18
validatesns==0.1.1
validators==0.14.2
vine==1.3.0
voluptuous==0.12.2
wcwidth==0.2.5
webassets==0.11
Werkzeug==0.12.2
xlrd==2.0.1
XlsxWriter==0.9.6
xlwt==1.3.0
zdesk==2.4.0
zeep==3.2.0
zenpy==2.0.12
zipp==1.2.0
zope.event==4.5.0
zope.interface==5.4.0
```

### How can we reproduce your problem?
Upgrading from ddtrace to version 0.58.4. The only dependency that changes during the upgrade is ddtrace.

### What is the result that you get?
Upgrading from ddtrace from version 0.57.3 to version 0.58.4 the number of open connections to our databases increased by 50% I spotted it easily with datadog itself. I think is related to the number of requests served by the service because we were not able to reproduce the error on our pre-production environment where the number of requests is much lower than production

### What is the result that you expected?
The number of connections should stay the same with both versions

### More info:
We deployed our application as a container running on Amazon ECS and the database is PostgreSQL version 13.3 database running on RDS.

**Comment by brettlangdon** [comment]
Just leaving some investigation notes here:

This is the diff between those two versions:

https://github.com/DataDog/dd-trace-py/compare/v0.57.3..v0.58.4

I don't see any changes to `django`, `psycopg2`, or `dbapi2` integrations... which is where I'd assume the source of the issue came from.


@Giaco9 would you be able to confirm for me if rolling back to `ddtrace==0.57.3` resolves the issue? If it does resolve it, then would you be willing to open a support case with support@datadoghq.com in order to continue investigating we will need to take a look at your account.

**Comment by Giaco9** [comment]
Hi @brettlangdon,
 sorry for the late reply. Rolling back to 0.57.4 solved the issue. I can open the support case but consider that we are terminating the trial so the account will be probably closed. Here is a screenshot of the graph I used to detect the problem
![Screenshot 2022-02-22 at 19 11 25](https://user-images.githubusercontent.com/4239892/155193311-091e7ced-104c-4c0c-9dfa-1075552a06d4.png)
Also here is a screenshot from CloudWatch
![Screenshot 2022-02-22 at 19 13 14](https://user-images.githubusercontent.com/4239892/155193570-2a814769-6a00-4175-bf9f-22c4e6b714b6.png)
I don't understand why the number of connections is different but the increase is the same.

**Comment by Giaco9** [comment]
@brettlangdon I opened the support case, ID: 696325

**Comment by technopig** [comment]
@Giaco9 The number of connections is probably different because of the rollup applied in Datadog - in DD, your graph was showing the "average of reported values" - meaning the average of each reported connection count, rather than the sum (which is what CW appears to be showing). Hope that helps.

**Comment by Giaco9** [comment]
Yes after i deeper investigation I saw the difference!
Thank you for the clarification, @technopig
Now I'm investigating with the support team, I can back to you as soon as we found a solution

**Comment by Giaco9** [comment]
Hi @brettlangdon,
 I reported it in the internal DataDog support case, but I think it deserves to be reported also here in case anyone else is experiencing the same issue. After many attempts, we found out that the root cause of the problem is the integration with `gevent`. Either removing `gevent` from our requirements or adding the environment variable `DD_TRACE_GEVENT_ENABLED=false` solved the issue. It was happening only when the application started with uwsgi and was not happening when the application started as either a celery worker or a Django long-running command. We are not sure which changes in version 0.58.0 caused the issue but for sure `gevent` is involved, also considering that something changed about `gevent` in version 0.58.0.
I would like to say thank you to the Datadog support team for their support! Thank you all, really great job!
